docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,129 @@
+- `api-default`: The defaults provided for omitted/nil fields by the API server.
+- `on-disk-config`: `KubeletConfiguration` provided on disk, used as a fallback when the Kubelet cannot cannot contact the API server.
+- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via `on-disk-config`.
+- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
+- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.
+
+### Well-defined configmap type for Kubelet configuration
+
+- There is an alpha version of the `KubeletConfiguration` type: external in `pkg/componentconfig/v1alpha1/types.go`, internal in `pkg/componentconfig/types.go`.
+- We likely want to group related fields of the `KubeletConfiguration` into substructures. Please note that `v1alpha1` will not be regrouped, but eventually we will start working on this.
+
+@timstclair and I did some brainstorming and came up with some rough potential categories:
+
+---
+- Networking
+- Runtimes (only one of these, in addition to Common, should be specified for a given Kubelet)
+    + Common
+    + Rkt
+    + Docker
+- Events
+- Security/Policies
+- Images (might be better as a runtime sub-category)
+- Resource Management
+    + GC/Eviction
+    + Allocatable/SystemReserved/KubeReserved
+    + Custom resources? e.g. NvidiaGPUs?
+- Cloud-provider-specific settings
+- Cluster/API
+    + KubeAPIQPS
+    + ContentType
+- SystemSpecification (this might be a non-category, all of the available hardware should theoretically be detectable by the Kubelet)
+    + NodeIP (probably belongs in networking)
+    + NvidiaGPUs could be here, but shouldn't the Kubelet be able to detect this?
+- Experimental/Volatile Features (could be tricky, might be better off using alpha versioning to separate experiments from the stable API)
+
+---
+
+### Bringing up a master node
+
+- Start the Kubelet with `minimal-config`
+- On-disk static pod manifests are used to bring up the API server, scheduler, controller-manager, etc.
+Once the master is ready, the user creates some configs via the API:
+- They might create a `global-kubelet-configuration`
+- They might create a `<node-name>-kubelet-configuration`, e.g. if they want a master-specific configuration.
+- Then the Kubelets' configuration sync loops will see the new config, and the Kubelets will restart to take up that new config.
+
+### Bringing up a Kubelet
+
+- Start the Kubelet with `minimal-config`
+- If `<current-node-name>-kubelet-config` exists on the API server, download and use the `<current-node-name>-kubelet-config`
```

Comment:
Technically, yes at the moment. I haven't thought very much about run-once Kubelet, and I need to learn more about its actual use cases before I can say whether this is the right thing to do. It might make sense to try to pull the config from the API server if run-once, but not start the configuration sync loop that kills the Kubelet if new config is detected. But I need to learn more before I can say for sure. 

What are some good examples of when --runonce is used?